pkg/odo/cli/list/list.go

```diff
@@ -0,0 +1,310 @@
+	// Do this before anything else, or else we will just error out with the:
+	// invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
+	// instead
+	if !dfutil.CheckKubeConfigExist() {
+		return errors.New("KUBECONFIG not found. Unable to retrieve cluster information. Please set your Kubernetes configuration via KUBECONFIG env variable or ~/.kube/config")
+	}
+
+	// Create the local context and initial Kubernetes client configuration
+	lo.Context, err = genericclioptions.New(genericclioptions.NewCreateParameters(cmdline))
+	if err != nil {
+		return err
+	}
+
+	// Check for the Devfile and then retrieve all information regarding the local Devfile
+	lo.devfilePath = location.DevfileLocation("")
+	if util.CheckPathExists(lo.devfilePath) {
+
+		// Parse the devfile
+		devObj, parseErr := devfile.ParseAndValidateFromFile(lo.devfilePath)
+		if parseErr != nil {
+			return parseErr
+		}
+
+		// Convert the devfile to a Component{} for table output and manipulation
+		// we want to limit using lo.EnvSpecificInfo, but fortunately, it will simply try to find
+		// lo.devfilePath for our usage.
+		localComponent, err := component.GetComponentFromDevfile(devObj)
+		if err != nil {
+			return err
+		}
+		lo.localComponent = localComponent
+
+		// Get global variables from the devfile
+		lo.componentType = component.GetComponentTypeFromDevfileMetadata(devObj.Data.GetMetadata())
+		lo.componentName = component.GetNameFromDevfileMetadata(devObj.Data.GetMetadata())
+
+		// Set the local project/namespace it's suppose to deploy to.. we check the context first from the devfile.yaml
+		// and set that.
+		lo.project = lo.Context.GetProject()
+
+		// Set the "local component" namespace since for table output
+		lo.localComponent.Namespace = lo.project
+
+		// We assume NOT PUSHED until we check the Kubernetes cluster
+		lo.localComponent.Status.State = component.StateTypeNotPushed
+
+	}
+
+	// If the context is "" (devfile.yaml not found..), we get the active one from KUBECONFIG.
+	if lo.project == "" {
+		lo.project = lo.KClient.GetCurrentNamespace()
+	}
+
+	// If the namespace flag has been passed, we will search there.
+	// if it hasn't, we will search from the default project / namespace.
+	if lo.namespaceFlag != "" {
+		lo.namespaceFilter = lo.namespaceFlag
+	} else {
+		lo.namespaceFilter = lo.project
+	}
+
+	return
+}
+
+// Validate ...
+func (lo *ListOptions) Validate() (err error) {
+	return nil
+}
+
+// Run has the logic to perform the required actions as part of command
+func (lo *ListOptions) Run() error {
+
+	// Set the correct selector / project before continuing in case we have passed in
+	// --namespace
+
+	// All variables pertaining to output / combination
+	devfileComponents := []component.Component{}
+
+	// Step 1.
+	// Retrieve all related components from the Kubernetes cluster, from the given namespace
+	if lo.KClient != nil {
```

Comment:
We want to get rid of the KClient from the context, please use the dependency injection instead:

1. Add `clientset.Add(componentListCmd, clientset.KUBERNETES)` in `NewCmdList`
2. Add a `clientset Clientset` field in the `ListOptions` structure
3. Add `o.clientset = clientset` in `SetClientset`
4. You can now use `o.clientset.KubernetesClient` instead of `o.KClient`